// react-fragment/index.js
/**
 * @providesModule Modal
 * @flow
 */
'use strict';

const Platform = require('Platform');
const PropTypes = require('react/lib/ReactPropTypes');
const React = require('React');
const StyleSheet = require('StyleSheet');
const UIManager = require('UIManager');
const View = require('View');
const deprecatedPropType = require('deprecatedPropType');

const requireNativeComponent = require('requireNativeComponent');
const RCTHalfCardView = requireNativeComponent('RCTHalfCardView', null);

/**
 * The HalfCard component is a simple way to present content above an enclosing view.
 *
 * _Note: If you need more control over how to present halfcards over the rest of your app,
 * then consider using a top-level Navigator. Go [here](/react-native/docs/navigator-comparison.html) to compare navigation options._
 *
 * ```javascript
 * import React, { Component } from 'react';
 * import { HalfCard, Text, TouchableHighlight, View } from 'react-native';
 *
 * class HalfCardExample extends Component {
 *
 *   constructor(props) {
 *     super(props);
 *     this.state = {halfcardVisible: false};
 *   }
 *
 *   setHalfCardVisible(visible) {
 *     this.setState({halfcardVisible: visible});
 *   }
 *
 *   render() {
 *     return (
 *       <View style={{marginTop: 22}}>
 *         <HalfCard
 *           visible={this.state.halfcardVisible}
 *           onRequestClose={() => {alert("HalfCard has been closed.")}}
 *           >
 *          <View style={{marginTop: 22}}>
 *           <View>
 *             <Text>Hello World!</Text>
 *
 *             <TouchableHighlight onPress={() => {
 *               this.setHalfCardVisible(!this.state.halfcardVisible)
 *             }}>
 *               <Text>Hide HalfCard</Text>
 *             </TouchableHighlight>
 *
 *           </View>
 *          </View>
 *         </HalfCard>
 *
 *         <TouchableHighlight onPress={() => {
 *           this.setHalfCardVisible(true)
 *         }}>
 *           <Text>Show HalfCard</Text>
 *         </TouchableHighlight>
 *
 *       </View>
 *     );
 *   }
 * }
 * ```
 */
class HalfCard extends React.Component {
  static propTypes = {
    visible: PropTypes.bool,
    onRequestClose: Platform.OS === 'android' ? PropTypes.func.isRequired : PropTypes.func,
    onShow: PropTypes.func,
  };

  static defaultProps = {
    visible: true,
  };

  render(): ?ReactElement<any> {
    if (this.props.visible === false) {
      return null;
    }

    return (
      <RCTHalfCardView
        onRequestClose={this.props.onRequestClose}
        onShow={this.props.onShow}
        style={styles.halfcard}
        onStartShouldSetResponder={this._shouldSetResponder}
        >
        <View>
          {this.props.children}
        </View>
      </RCTHalfCardView>
    );
  }

  _shouldSetResponder(): boolean {
    return true;
  }
}

module.exports = HalfCard;
